peps/pep-0667.rst

```diff
@@ -201,15 +413,222 @@ will continue to work correctly. Debuggers and other tools that use
 even in the presence of threaded code, coroutines and generators.
 
-C-API
------
+``frame.f_locals`` compatibility
+--------------------------------
+
+Although ``f.f_locals`` behaves as if it were the namespace of the function,
+there will be some observable differences.
+For example, ``f.f_locals is f.f_locals`` will be ``False`` for optimized
+frames, as each access to the attribute produces a new write-through proxy
+instance.
+
+However ``f.f_locals == f.f_locals`` will be ``True``, and
+all changes to the underlying variables, by any means, including the
+addition of new variable names as mapping keys, will always be visible.
+
+``locals()`` compatibility
+''''''''''''''''''''''''''
+
+``locals() is locals()`` will be ``False`` for optimized frames, so
+code like the following will raise ``KeyError`` instead of returning
+``1``::
+
+    def f():
+        locals()["x"] = 1
+        return locals()["x"]
+
+To continue working, such code will need to explicitly store the namespace
+to be modified in a local variable, rather than relying on the previous
+implicit caching on the frame object::
+
+    def f():
+        ns = {}
+        ns["x"] = 1
+        return ns["x"]
+
+While this technically isn't a formal backwards compatibility break
+(since the behaviour of writing back to ``locals()`` was explicitly
+documented as undefined), there is definitely some code that relies
+on the existing behaviour. Accordingly, the updated behaviour will
+be explicitly noted in the documentation as a change and it will be
+covered in the Python 3.13 porting guide.
+
+To work with a copy of ``locals()`` on all versions without making
+redundant copies on Python 3.13+, a helper function will need to
+be defined in the affected code::
+
+    if sys.version_info >= (3, 13):
+        def _ensure_func_snapshot(d):
+            return d # 3.13+ locals() already returns a snapshot
+    else:
+        def _ensure_func_snapshot(d):
+            return dict(d) # Create snapshot on older versions
+
+    def f():
+        ns = _ensure_func_snapshot(locals())
+        ns["x"] = 1
+        return ns
+
+In other scopes, ``locals().copy()`` can continue to be called
+unconditionally without introducing any redundant copies.
+
+Impact on ``exec()`` and ``eval()``
+'''''''''''''''''''''''''''''''''''
+
+Even though this PEP does not modify ``exec()`` or ``eval()`` directly,
+the semantic change to ``locals()`` impacts the behavior of ``exec()``
+and ``eval()`` as they default to running code in the calling namespace.
+
+While the exact wording in the library reference is not entirely explicit,
+both ``exec()`` and ``eval()`` have long used the results of calling
+``globals()`` and ``locals()`` in the calling Python frame as their default
+execution namespace.
+
+This was historically also equivalent to using the calling frame's
+``frame.f_globals`` and ``frame.f_locals`` attributes, but this PEP maps
+them to ``globals()`` and ``locals()`` in order to preserve the property
```

Comment:
"them" is ambiguous.  Does it refer to `frame.f_globals` and `frame.f_locals` or the arguments to `exec()` and `eval()`?